### PR TITLE
Adding complex number support for all_to_all/scatter

### PR DESCRIPTION
**Pull request by marjanfariborz**
Summary: Modifying all_to_all and scatter to support complex numbers as well as float numbers.

Test Plan: buck run //caffe2/test/distributed:distributed_gloo_fork -- test_name --print-passing-details --run-disabled

Differential Revision: D29563938

### CODE REVIEW

File: torch/distributed/distributed_c10d.py [hunk 1]
```diff
@@ -2179,6 +2179,8 @@ def scatter(tensor, scatter_list=None, src=0, group=None, async_op=False):
     """
     Scatters a list of tensors to all processes in a group.
 
+    Complex tensors are supported.
```
Comment (wanchaol) [inline torch/distributed/distributed_c10d.py:2182]: nit: can we make this right before the `Args` section? we don't want to separate the descriptions

File: torch/distributed/distributed_c10d.py [hunk 2]
```diff
@@ -2453,13 +2459,35 @@ def all_to_all_single(
...
+        >>> input
+        tensor([1+1j, 2+2j, 3+3j, 4+4j])     # Rank 0
+        tensor([5+5j, 6+6j, 7+7j, 8+8j])     # Rank 1
+        tensor([9+9j, 10+10j, 11+11j, 12+12j])   # Rank 2
+        tensor([13+13j, 14+14j, 15+15j, 16+16j]) # Rank 3
+        >>> output = torch.empty([4], dtype=torch.int64)
+        >>> dist.all_to_all_single(output, input)
+        >>> output
+        tensor([1+1j, 5+5j, 9+9j, 13+13j])    # Rank 0
+        tensor([2+2j, 6+6j, 10+10j, 14+14j])    # Rank 1
```
Comment (wanchaol) [inline torch/distributed/distributed_c10d.py:2475]: nit: lint the trailing comment

File: torch/distributed/distributed_c10d.py [hunk 3]
```diff
@@ -2549,6 +2579,23 @@ def all_to_all(output_tensor_list, input_tensor_list, group=None, async_op=False
...
+        >>> input
+        [tensor([1+1j]), tensor([2+2j]), tensor([3+3j]), tensor([4+4j])]     # Rank 0
+        [tensor([5+5j]), tensor([6+6j]), tensor([7+7j]), tensor([8+8j])]     # Rank 1
+        [tensor([9+9j]), tensor([10+10j]), tensor([11+11j]), tensor([12+12j])]   # Rank 2
+        [tensor([13+13j]), tensor([14+14j]), tensor([15+15j]), tensor([16+16j])] # Rank 3
+        >>> output = list(torch.empty([4], dtype=torch.int64).chunk(4))
+        >>> dist.all_to_all(output, input)
+        >>> output
+        [tensor([1+1j]), tensor([5+5j]), tensor([9+9j]), tensor([13+13j])]    # Rank 0
+        [tensor([2+2j]), tensor([6+6j]), tensor([10+10j]), tensor([14+14j])]    # Rank 1
```
Comment (wanchaol) [inline torch/distributed/distributed_c10d.py:2595]: ditto: lint the comment

File: torch/testing/_internal/distributed/distributed_test.py [hunk 4]
```diff
@@ -2670,6 +2674,11 @@ def test_scatter(self):
             group, group_id, rank = self._init_global_test()
             self._test_scatter_helper(group, group_id, rank)
 
+        @unittest.skipIf(BACKEND == "nccl", "Nccl does not support scatter")
+        def test_scatter_complex(self):
+            group, group_id, rank = self._init_global_test()
+            self._test_scatter_helper(group, group_id, rank, dtype=torch.cfloat)
```
Comment (wanchaol) [inline torch/testing/_internal/distributed/distributed_test.py:2680]: what's the difference between `torch.cfloat` and `torch.complex64`?

**Comment by wanchaol** [comment]
Looks good overall 👍 , thanks for working on this! Just have some minor comments/questions. For CI errors, I checked them and they seems not related to your change, so we are good to ship once we address the comments.

File: torch/testing/_internal/distributed/distributed_test.py (hunk 4 above)
Comment (marjanfariborz) [reply torch/testing/_internal/distributed/distributed_test.py:2680]: They are both representing 64-bit complex data type.

File: torch/distributed/distributed_c10d.py (hunk 1 above)
Comment (marjanfariborz) [reply torch/distributed/distributed_c10d.py:2182]: Sure thing.

**Review by rohan-varma — APPROVED** [review]
Awesome job and very comprehensive testing! Please also get a stamp from @wanchaol

**Review by wanchaol — APPROVED** [review]

**Pull request closed without merging**